Prometheus: Interval/Step changes

torkelo opened the pull request:
This PR / Branch Tries to make Prometheus more like the other data sources in terms of interval handling. 

This means that you can now set a interval lower limit on panel level (all queries), the panel has a `$interval` variable by default (meaning you can use this variable in your query even if you do not have a variable named interval). 

If you have a template variable named `interval` then that will be used. 

![image](https://cloud.githubusercontent.com/assets/10999/13534867/84e4fad6-e237-11e5-8108-3d3956a8ebbc.png)

Other changes:
- removed resolution, this parameter was very confusing as it override the step parameter and was just strange. 

TODO:
- fix alignment of time range to interval

torkelo committed:
feat(prometheus): work on changing on prometheus interval / step is used and can be used in the query

torkelo commented:
@mtanda @grobie thoughts?

grobie commented:
Thanks! On my review queue, will comment latest tomorrow.

mtanda commented on public/app/plugins/datasource/prometheus/datasource.ts:
```diff
@@ -62,28 +61,27 @@ export function PrometheusDatasource(instanceSettings, $q, backendSrv, templateS
-    _.each(options.targets, _.bind(function(target) {
+    var queries = _.map(options.targets, target => {
       if (!target.expr || target.hide) {
         return;
```
If use map(), need to filter invalid query generated at this line.

mtanda commented on public/app/plugins/datasource/prometheus/datasource.ts:
```diff
@@ -62,28 +61,27 @@ export function PrometheusDatasource(instanceSettings, $q, backendSrv, templateS
-      if (query.step !== 0 && range / query.step > 11000) {
-        target.step = query.step = Math.ceil(range / 11000);
+      if (range / query.step > 11000) {
+        query.step = Math.ceil(range / 11000);
```
not sure. the value of "interval" need to be updated here?

mtanda replied:
@torkelo Please let me do the supplementary explanation of this comment.
I think, the plugin should re-calculate "interval" from "step" at the line.

Before this line, the "step" is calculated from "interval", these values are syncrhonized. (If "interval" is "10s", the "step" is "10", right?)
But, after this line, these values could be not synchronized. ("interval" is "10s", but the "step" become like "50")

Refer comment from original PR, https://github.com/grafana/grafana/pull/3757#issuecomment-191713400,
> Now say the step is auto-calcualted as 1h (on longer periods).
Every 1h datapoint is calculated as avg from values of now() and now()-5m. It's not correct, it should be of now() and now()-1h.

He want to sync the "interval" and "step" after the "step" is auto-calculated.

So, I think, the plugin should re-calculate synced "interval" from "step".

mtanda commented:
Thanks!
I think, this is good approach to fix original issue.

@torkelo Would you merge below line to this PR?
https://github.com/grafana/grafana/pull/3757/files#diff-8febf6634a5b868cd51400d863e1df45R66

torkelo committed:
feat(prometheus): minor changes from #3757

torkelo commented:
@mtanda I updated with one change from #3757 but the second change you pointed to does not make sense, there is no step property on the query model, the only reason that could have worked was that before the prometheus datasource set this property here (https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/prometheus/datasource.ts#L77) 

But that is not correct usage of the query model so I removed that. The query function should not update the query mode with a time range dependent value that changes.

mtanda commented:
@torkelo Sorry, as you pointed out, I think updating query model is bad code.

The code is appeared in this PR, https://github.com/grafana/grafana/pull/3284.
To fill null at gap point, the auto-calculated "step" is needed, and to pass "step" to "transformMetricData()", I update query model at the line (https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/prometheus/datasource.ts#L77).

The "step" is referred at this line.
https://github.com/grafana/grafana/blob/277d200bdb2f121aa8f0bca8d3ee2fb6b5c8bbd6/public/app/plugins/datasource/prometheus/datasource.ts#L219

I should avoid updating query model, I should change the function parameter like,
```
this.transformMetricData = function(md, options, start, end, step)
```
and pass "step" as another parameter.

grobie commented:
I tried to test this, but I wasn't able to execute any query so far, the step parameter in the generated request is always `undefined`. It doesn't help to set either of the two interval fields.

General comments (based on my current limited understanding of that change):
* Always having an `$interval` variable available is awesome! This helps a lot to create rate-per-$interval queries in combination with the increase() function.
* I've actually used the resolution parameter quite a few times so far, as it helps to reduce the query execution times for expensive expressions. While I agree that it might be confusing for new users why there are two parameters, I understand the step parameter to set an absolute value and the resolution parameter to set a relative value. E.g. even when zooming out, don't ask for too many data points.
* I don't understand the use case for the second interval input field. From looking at our usage, graphs with more than one query are the small minority. I'd assume it adds more confusion than benefit by only having it to set once. Also being able to set a lower absolute step is not too helpful to prevent queries from requesting too many data points. If I zoom out and look at a date range of days, it's helpful to limit the maximum number of data points fetched from Prometheus. With a setting of `>1m` it's still possible to request thousands of data points when looking at long date ranges. That's what the resolution parameter in promdash and current grafana is being used for by us.
* IMHO `legend` describes the use case more accurately than `alias`, but I assume that rename was made to consolidate the naming with graphite's `alias()` function?
* In my chromium/arch browser, there is a line break in the alias pattern tooltip: 
![screenshot from 2016-03-07 00-20-12](https://cloud.githubusercontent.com/assets/3432/13561395/a1a90bf2-e3fa-11e5-9764-9a899cd6ad0f.png)

torkelo committed:
fix(prometheus): fixed issue with recent commit

torkelo commented:
@grobie thanks for reviewing! 

>I tried to test this, but I wasn't able to execute any query so far, the step parameter in the generated request is always undefined. It doesn't help to set either of the two interval fields.

1. Sorry, I accidentally broke it yesterday evening, fixed now.

>I've actually used the resolution parameter quite a few times so far, as it helps to reduce the query execution times for expensive expressions. While I agree that it might be confusing for new users why there are two parameters, I understand the step parameter to set an absolute value and the resolution parameter to set a relative value. E.g. even when zooming out, don't ask for too many data points.

2. The resolution setting was pretty confusing as it actually changed the setting before it (the interval), so you had two options that control the same property. I do not understand why this setting makes prometheus queries faster, Prometheus still has to process all metrics to do the query? And returning 500 points instead of 200 should make no big difference. 

>I don't understand the use case for the second interval input field. ...

3. The second interval field is a little confusing I agree, it uses a feature that other data sources are using (InfluxDB), it allows you to set an lower interval limit for all queries (it's a property set on the panel not on a query), it allows the panel to do some interval checking (and maybe align panel time range to interval in a future change). We can remove this option and have the interval on all queries use the panel interval (bind to the same property, but then queries with different intervals could not be issued ont he same graph)

>MHO legend describes the use case more acc...

* The reason for legend format -> Alias is that series names are referred to as alias in other places (series overrides), I agree it is not the most intuitive name for "series name" 

>In my chromium/arch browser, there is a line break in the alias pattern tooltip:
* Strange that it looks like that in Chromium for you, looks fine here. Will try fix.

torkelo committed:
feat(prometheus): form fixes

roman-vynar commented:
Thanks guys for making this a hot topic 😉

grobie commented:
> The resolution setting was pretty confusing as it actually changed the setting before it (the interval), so you had two options that control the same property. I do not understand why this setting makes prometheus queries faster, Prometheus still has to process all metrics to do the query? And returning 500 points instead of 200 should make no big difference.

The execution model of Prometheus means a given expression is evaluated at any requested timestamp (the data is only loaded once from disk if it's not in memory already). Let's say the time range is 1 day, then an interval of 1m means the expression needs to be evaluated 1440 times, with an interval of 2m it's only 720 times. Here a practical example:

```
# setup
Metric: haproxy_backend_connections_total
Number of timeseries: ~1500 (main labels: HAProxy instance, backend name)
Expression: topk(5, sum by(backend)(rate(haproxy_backend_connections_total[1m])))

# evaluation times
Resolution 30s: 24947ms
Resolution 60s: 12023ms
Resolution 120s: 6901ms
Resolution 180s: 3318ms
```

Once data is loaded from disk, resolution linearly influences execution time. Keep in mind that Prometheus is still a fairly young database and will show a lot of improvements here in the future, but that's where we are right now.

The above explanation hopefully shows why having the ability to influence the resolution is very helpful, besides choosing a good default. Looking at the current auto calculated value, I'd vote to reduce the resolution by half.

I wonder if we can just always interpret the set interval to be relative to the time range. E.g. manually setting interval to 1m when looking at a 1h range, will result in a interval of 24m when looking at a 24h range.

> The second interval field is a little confusing I agree, it uses a feature that other data sources are using (InfluxDB), it allows you to set an lower interval limit for all queries (it's a property set on the panel not on a query), it allows the panel to do some interval checking (and maybe align panel time range to interval in a future change). We can remove this option and have the interval on all queries use the panel interval (bind to the same property, but then queries with different intervals could not be issued ont he same graph)

Only having one way to set the interval sounds favorable to me.

I currently have a single case with two queries in one panel with different intervals (trading resolution for speed in the expensive expression, while keeping a high resolution in the cheap expression). Would it be possible to keep that behavior while still not introducing the panel interval field?

roman-vynar commented:
I still hope this will be included into Grafana 3.0.0.

By the way, another improvement: if the period chosen is too small and "auto" interval is auto-defined as "Auto interval step" of 200, it may result to `ms` unit which is not supported by Prometheus.

For example, Grafana may generate:
```
rate(node_vmstat_pgpgin[100ms])
```
However, there is no `ms` in Prometheus resulting an error.

jemc commented:
@torkelo - any more progress on this one?

jemc commented:
Also, it's not clear to me whether this change also deals with offsets in the grafana time range?

torkelo commented:
@jemc no progress from my part, kind of wanted to kick start the improvement and have users who use and know prometheus better test and complete this. but so far there hasn't been much happening

roman-vynar commented:
Using Grafana 3.0.1. No changes in these regards.

Can we simply make it so the `Step` field supports a template variable? It's one line change and will align `rate()` interval with the actual datapoint step.

Other issues: `ms` should not be passed to Prometheus being unsupported unit; Prometheus Link should pass `Step` and populate template variables w/o surrounding `{` `}`.

Let me know if I can help.

acidmoose commented:
Letting Step support a template variable would also fix my current issue with regards to zooming on a delta() or increase() function as well. Zooming and then changing the step to equal the selected range-vector template yields perfect graphs. This is especially noticeable when using bars for graphs.

torkelo commented:
closing this, the prometheus data source needs to big refactortings, pretty messy right now, hopefully I or someone from the prometheus community will find the time :)

roman-vynar commented:
We have the template var support in the `Step` field with Grafana 4.0, so it is perfect anyway. Thanks!
